### PR TITLE
s3 adapter multi region

### PR DESCRIPTION
**Pull request by johnnyaug**
Solves #2245, through #2325 + #2326.

The new solution (lazily) creates an S3 client for every AWS region.
The s3 block adapter will use the suitable S3 client according to the region of the underlying bucket.

When the underlying storage is an S3-compatible service, with no region support (e.g. MinIO), a default S3 client is created and no change in behavior is expected

### CODE REVIEW

File: .github/workflows/nessie.yaml [hunk 1]
```diff
@@ -129,18 +129,18 @@ jobs:
...
         with:
-          registry: docker.pkg.github.com
+          registry: ghcr.io
           username: ${{ github.actor }}
           password: ${{ secrets.GITHUB_TOKEN }}
       - name: Pull last built docker image cache
         run: |
-          docker pull docker.pkg.github.com/${GITHUB_REPOSITORY,,}/build-cache:${{ steps.version.outputs.tag }} || \
-          docker pull docker.pkg.github.com/${GITHUB_REPOSITORY,,}/build-cache || \
+          docker pull ghcr.io/${GITHUB_REPOSITORY,,}/build-cache:${{ steps.version.outputs.tag }} || \
```
Comment (johnnyaug) [inline .github/workflows/nessie.yaml:137]: reviewers - please ignore the ghcr.io change, it will be introduced in another PR.

File: pkg/block/s3/inventory.go [hunk 2]
```diff
@@ -23,35 +22,35 @@ type Manifest struct {
...
-type inventoryFile struct {
+type InventoryFile struct {
 	Key      string `json:"key"` // an s3 key for an inventory list file
 	firstKey string
 	lastKey  string
 }
 
 func (a *Adapter) GenerateInventory(ctx context.Context, logger logging.Logger, manifestURL string, shouldSort bool, prefixes []string) (block.Inventory, error) {
-	return GenerateInventory(logger, manifestURL, a.s3, s3inventory.NewReader(ctx, a.s3, logger), shouldSort, prefixes)
+	m, err := a.loadManifest(ctx, manifestURL)
```
Comment (talSofer) [inline pkg/block/s3/inventory.go:38]: As discussed, we shall open an issue to adding import integration tests.

File: pkg/block/s3/adapter.go [hunk 3]
```diff
@@ -109,6 +116,29 @@ func NewAdapter(s3 s3iface.S3API, opts ...func(a *Adapter)) *Adapter {
...
+	region, err := s3manager.GetBucketRegion(ctx, a.awsSession, bucket, "")
+	if err != nil {
+		a.log(ctx).WithError(err).Error("failed to get region for bucket, falling back to default region")
+		region = *a.awsSession.Config.Region
+	}
+	a.regionByBucket[bucket] = region
+	return region
+}
+
+func (a *Adapter) s3Client(ctx context.Context, bucket string) s3iface.S3API {
```
Comment (talSofer) [inline pkg/block/s3/adapter.go:133]: are you planning to add unit tests for this piece? I mean to test that clients are created lazily.

File: pkg/block/s3/adapter.go [hunk 4]
```diff
@@ -62,7 +65,9 @@ func resolveNamespacePrefix(opts block.WalkOpts) (block.QualifiedPrefix, error)
 }
 
 type Adapter struct {
-	s3                    s3iface.S3API
+	awsSession            *session.Session
+	s3ClientByRegion      map[string]s3iface.S3API
+	regionByBucket        map[string]string
```
[the remark this answers is not in the document]
Comment (johnnyaug) [reply pkg/block/s3/adapter.go:70]: These are bucket names, not storage namespaces. The confusion stems from bad naming we have on the `QualifiedKey` struct. In the context of this struct, "StorageNamespace" is actually the bucket name.

File: pkg/block/s3/adapter.go [hunk 5]
```diff
@@ -62,7 +65,9 @@ func resolveNamespacePrefix(opts block.WalkOpts) (block.QualifiedPrefix, error)
 }
 
 type Adapter struct {
-	s3                    s3iface.S3API
+	awsSession            *session.Session
+	s3ClientByRegion      map[string]s3iface.S3API
```
[the remark this answers is not in the document]
Comment (johnnyaug) [reply pkg/block/s3/adapter.go]: Done

File: pkg/block/s3/inventory.go (hunk 2 above)
Comment (johnnyaug) [reply pkg/block/s3/inventory.go:38]: Opened #2332

File: pkg/block/s3/adapter.go (hunk 3 above)
Comment (johnnyaug) [reply pkg/block/s3/adapter.go:133]: On it

File: pkg/block/s3/adapter.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (talSofer) [reply pkg/block/s3/adapter.go]: Can you rename the field in QualifiedKey to bucket name or open an issue to keep track of that?

File: pkg/block/s3/adapter.go [hunk 6]
```diff
@@ -109,6 +116,29 @@ func NewAdapter(s3 s3iface.S3API, opts ...func(a *Adapter)) *Adapter {
 	return a
 }
 
+func (a *Adapter) getBucketRegion(ctx context.Context, bucket string) string {
```
Comment (itaiad200) [inline pkg/block/s3/adapter.go:119]: This isn't thread safe, is it possible to reach here from different concurrent paths on startup? Also, the locking will also save you from the need to call s3 for the same request over and over again.

[the remark this answers is not in the document]
Comment (itaiad200) [reply pkg/block/s3/adapter.go:119]: And then what happens if there are two storage namespaces within the same bucket? 🤯

File: pkg/block/s3/adapter.go [hunk 7]
```diff
@@ -109,6 +116,29 @@ func NewAdapter(s3 s3iface.S3API, opts ...func(a *Adapter)) *Adapter {
...
+		a.log(ctx).WithError(err).Error("failed to get region for bucket, falling back to default region")
+		region = *a.awsSession.Config.Region
+	}
+	a.bucketToRegion[bucket] = region
+	return region
+}
+
+func (a *Adapter) s3Client(ctx context.Context, bucket string) s3iface.S3API {
+	region := a.getBucketRegion(ctx, bucket)
+	if _, hasClient := a.regionToS3Client[region]; !hasClient {
```
Comment (itaiad200) [inline pkg/block/s3/adapter.go:135]: Thread safety?

File: pkg/block/s3/inventory.go [hunk 8]
```diff
@@ -23,35 +22,35 @@ type Manifest struct {
...
+	if err != nil {
+		return nil, err
+	}
+	svc := a.s3Client(ctx, m.inventoryBucket)
+	return GenerateInventory(logger, m, s3inventory.NewReader(ctx, svc, logger), shouldSort, prefixes)
 }
 
-func GenerateInventory(logger logging.Logger, manifestURL string, s3 s3iface.S3API, inventoryReader s3inventory.IReader, shouldSort bool, prefixes []string) (block.Inventory, error) {
+func GenerateInventory(logger logging.Logger, m *Manifest, inventoryReader s3inventory.IReader, shouldSort bool, prefixes []string) (block.Inventory, error) {
```
Comment (itaiad200) [inline pkg/block/s3/inventory.go:46]: Should still be exported?

File: pkg/block/s3/adapter.go [hunk 9]
```diff
@@ -126,7 +156,7 @@ func (a *Adapter) Put(ctx context.Context, obj block.ObjectPointer, sizeBytes in
 		Key:          aws.String(qualifiedKey.Key),
 		StorageClass: opts.StorageClass,
 	}
-	sdkRequest, _ := a.s3.PutObjectRequest(&putObject)
+	sdkRequest, _ := a.s3Client(ctx, qualifiedKey.StorageNamespace).PutObjectRequest(&putObject)
```
Comment (itaiad200) [inline pkg/block/s3/adapter.go:159]: `qualifiedKey.StorageNamespace` is sometimes different than the bucket.

**Review by itaiad200 — CHANGES_REQUESTED** [review]
Overall looks good. I think there are some issues related to thread safety.

Also - how will this change the collected metrics? I think we should at least add another key for the number of regions per installation.

**Review by itaiad200 — CHANGES_REQUESTED** [review]
Overall looks good. I think there are some issues related to thread safety.

Also - how will this change the collected metrics? I think we should at least add another key for the number of regions per installation.

File: pkg/block/s3/adapter.go (hunk 9 above)
Comment (johnnyaug) [reply pkg/block/s3/adapter.go:159]: Is it? It used as the bucket for the request 4 lines above

File: pkg/block/s3/inventory.go (hunk 8 above)
Comment (johnnyaug) [reply pkg/block/s3/inventory.go:46]: Yes, used in the test

File: pkg/block/s3/adapter.go (hunk 7 above)
Comment (johnnyaug) [reply pkg/block/s3/adapter.go]: Changed to sync.Map. Thanks!

File: pkg/block/s3/inventory.go (hunk 2 above)
Comment (talSofer) [reply pkg/block/s3/inventory.go:38]: Thanks!

**Review by talSofer — APPROVED** [review]
Looks good! 
I think it's worth mentioning in the issue description what's the expected behavior for s3-compatible objects storage (e.g., Minio)
